### PR TITLE
unsafe integer arithmetic implementation

### PR DESCRIPTION
**Pull request by guipublic**
We benefit the unsafe behavior to implement
the bit size modulo for integer arithmetic in Noir only when required
(greedy algo)

### CODE REVIEW

File: crates/acir/src/circuit/gate.rs [hunk 1]
```diff
@@ -43,7 +43,28 @@ impl Gate {
 #[derive(Clone, Debug)]
 /// Directives do not apply any constraints.
 pub enum Directive {
-    Invert { x: Witness, result: Witness },
+    Invert {
+        x: Witness,
+        result: Witness,
+    },
+    Quotient {
```
Comment (kevaundray) [inline crates/acir/src/circuit/gate.rs:50]: Can we add comments above these to specify what they are doing?

For example, for invert, we can say:

"Inverts the value of x and stores it in the result variable"

Comment (guipublic) [reply crates/acir/src/circuit/gate.rs:53]: Done

**Comment by guipublic** [comment]
tests are failing because this branch fixes the bug revealed by the test case '5_over', but this test case is not part of this branch.
we should set it to pass instead of fail and merge this branch. (Or exclude it, merge and then activate it).

**Review by kevaundray — CHANGES_REQUESTED** [review]

File: crates/noirc_evaluator/src/object/integer.rs [hunk 2]
```diff
@@ -35,13 +55,72 @@ impl Integer {
...
+            } else {
+                evaluator
+                    .gates
+                    .push(Gate::Range(self.witness, self.num_bits));
+            }
         }
         Ok(())
     }
 
+    pub fn truncate(&self, evaluator: &mut Evaluator) -> Result<Integer, RuntimeErrorKind> {
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/object/integer.rs:90]: Done

File: crates/noirc_evaluator/src/object/integer.rs [hunk 3]
```diff
@@ -82,13 +177,109 @@ impl Integer {
...
+        )?;
+
+        Integer::from_object_with_max_bits(res, self.num_bits, new_bits, evaluator)
+    }
+
+    //Truncate a and/or b in case of overflow
+    pub fn truncate_argments(
+        &self,
+        b: Integer,
+        evaluator: &mut Evaluator,
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/object/integer.rs:241]: Done

File: crates/noirc_evaluator/src/object/integer.rs [hunk 4]
```diff
@@ -81,13 +177,109 @@ impl Integer {
...
+        f: fn(u32, u32) -> u32,
+    ) -> Result<(u32, Witness, Witness), RuntimeErrorKind> {
+        let mut new_bits = f(self.max_bits, b.max_bits);
+        let mut a_wit = self.witness;
+        let mut b_wit = b.witness;
+        if new_bits >= FieldElement::max_num_bits() {
+            if self.max_bits > b.max_bits {
+                a_wit = self.truncate(evaluator).unwrap().witness;
+                new_bits = f(self.num_bits, b.max_bits);
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/object/integer.rs:250]: Done: result_bits

File: crates/noirc_evaluator/src/object/integer.rs [hunk 5]
```diff
@@ -160,6 +436,7 @@ impl Integer {
         Ok(Integer {
             witness: result,
             num_bits: self.num_bits,
+            max_bits: self.num_bits, //XXX ??
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/object/integer.rs]: Removed, the result of the logic gate has self.num_bits size

**Comment by kevaundray** [comment]
LGTM

**Review by kevaundray — APPROVED** [review]

**Pull request merged**